rustler/src/types/binary.rs

```diff
@@ -1,43 +1,105 @@
+//! Safe wrappers around Erlang binaries.
+//!
+//! Rustler provides two types for wrapping Erlang binaries: [`Binary`] and
+//! [`OwnedBinary`].  Both represent a contiguous region `u8`s, and they both use
+//! the Erlang allocator. The primary difference between the two is ownership
+//! semantics.
+//!
+//! The _owned_ in `OwnedBinary` refers to the fact that it owns the VM binary is
+//! wraps, and the _owner_ of an `OwnedBinary` is free to modify its contents
+//! without fear of access violations or upsetting the VM. Ownership lasts until it
+//! is dropped by either falling out of scope, or converting it into a regular
+//! `Binary` via [`to_owned`] or one of the conversion traits. Once dropped, the VM
+//! is allowed to the free the backing memory at its leisure. Similar to `Vec`,
+//! `OwnedBinary` does not implement `copy` [^1], so passing one around by value is
+//! cheap and does not incur any additional heap allocations.
+//!
+//! The `Binary` type is an immutable shared-reference to a VM binary. `Binary`s are
+//! cheap to copy: all copies of of `Binary` point to the original `Binary`'s
+//! data. Additionally, a `Binary`'s lifetime is tied to the lifetime of a VM
```

Comment:
The `Env` is not a VM. That makes it sound like you will get the same `Env` on every NIF call, which is not the case. The `Binary`'s lifetime, like all other VM-owned objects, is tied to exactly one NIF execution. You are not allowed to access any of these objects once the NIF is finished.

Reply:
Actually, the second part of the sentence makes that clear, but `Env != VM` is still important, I think.

Reply:
Actually actually, the `Env` is always either a "process environment" (i.e. tied to one particular process) or a manually allocated one to create terms to send around.